robot/impl/local_robot_test.go

```diff
@@ -2426,116 +2426,205 @@ func TestDependentResources(t *testing.T) {
+		test.That(t, res, test.ShouldBeNil)
+
+		// Remove module entirely.
+		cfg3 := &config.Config{
+			Components: []resource.Config{
+				{
+					Name:  "g",
+					Model: gizmoModel,
+					API:   gizmoAPI,
+				},
+			},
+			Services: []resource.Config{
+				{
+					Name:  "s",
+					Model: summationModel,
+					API:   summationAPI,
+				},
+			},
+		}
+		r.Reconfigure(ctx, cfg3)
 
-	// Remove module entirely.
-	cfg3 := &config.Config{
-		Components: []resource.Config{
-			{
-				Name:  "g",
-				Model: gizmoModel,
-				API:   gizmoAPI,
+		// Assert that adding module 'mod' back with original executable path re-adds
+		// modular component 'g' and modular service 's'.
+		r.Reconfigure(ctx, cfg)
+
+		_, err = r.ResourceByName(gizmoapi.Named("g"))
+		test.That(t, err, test.ShouldBeNil)
+		_, err = r.ResourceByName(summationapi.Named("s"))
+		test.That(t, err, test.ShouldBeNil)
+	})
+
+	t.Run("automatic reconfiguration", func(t *testing.T) {
+		cfg := &config.Config{
+			Modules: []config.Module{
+				{
+					Name:    "mod",
+					ExePath: rutils.ResolveFile("module/testmodule/testmodule"),
+				},
 			},
-		},
-		Services: []resource.Config{
-			{
-				Name:  "s",
-				Model: summationModel,
-				API:   summationAPI,
+			Components: []resource.Config{
+				{
+					Name:  "h",
+					Model: helperModel,
+					API:   generic.API,
+				},
 			},
-		},
-	}
-	r.Reconfigure(ctx, cfg3)
+		}
+		r.Reconfigure(ctx, cfg)
 
-	// Assert that adding module 'mod' back with original executable path re-adds
-	// modular component 'g' and modular service 's'.
-	r.Reconfigure(ctx, cfg)
+		h, err := r.ResourceByName(generic.Named("h"))
+		test.That(t, err, test.ShouldBeNil)
 
-	_, err = r.ResourceByName(gizmoapi.Named("g"))
-	test.That(t, err, test.ShouldBeNil)
-	_, err = r.ResourceByName(summationapi.Named("s"))
-	test.That(t, err, test.ShouldBeNil)
+		// Assert that removing testmodule binary and killing testmodule orphans
+		// helper 'h' a couple seconds after third restart attempt.
+		err = os.Remove(rutils.ResolveFile("module/testmodule/testmodule"))
+		test.That(t, err, test.ShouldBeNil)
+		_, err = h.DoCommand(ctx, map[string]interface{}{"command": "kill_module"})
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err.Error(), test.ShouldContainSubstring,
+			"error reading from server")
+
+		// Wait for "attempt 3" in logs.
+		testutils.WaitForAssertionWithSleep(t, time.Second, 20, func(tb testing.TB) {
+			tb.Helper()
+			test.That(tb, logs.FilterMessageSnippet("attempt 3").Len(),
+				test.ShouldEqual, 1)
+		})
+		time.Sleep(2 * time.Second)
+
+		_, err = r.ResourceByName(generic.Named("h"))
+		test.That(t, err, test.ShouldNotBeNil)
+		test.That(t, err, test.ShouldBeError,
+			resource.NewNotFoundError(generic.Named("h")))
+
+		// Assert that recompiling testmodule, removing testmodule and 'h' from
+		// config and adding both back re-adds 'h'.
+		//
+		// TODO(RSDK-2876): assert that we can keep 'h' in the config and it gets
```

Comment:
We may have to wait on RSDK-2876 for asserting something like
```
cfg2 := &config.Config {
   Components: []resource.Config{
      {
         Name: "h",
         Model: helperModel,
         API: generic.API
      }
   }
}
r.Reconfigure(ctx, &config.Config{})
r.Reconfigure(ctx, cfg)

h, err = r.ResourceByName(generic.Named("h"))
test.That(t, err, test.ShouldBeNil)
```
For the above assertions to pass, `removeOrphanedResources` would have to do something like [this logic](https://github.com/viamrobotics/rdk/blob/main/robot/impl/local_robot.go#L959-L978), but it can't because it has no access to the local robot's stored config. Rather than hacking that in, it may be wiser to just implement RSDK-2876 (probably my next ticket).

Reply:
sgtm